### ISSUE TITLE
Export to csv doesn't apply angular cellfilters (eg date format)

### ISSUE THREAD

**Issue by lexerka** [angular-ui/ui-grid#4214]
We have cellFilter applied (eg. to show time in HH:mm format) in the UI grid but when we export data into csv the filter is not applied which gives us something different then is shown on UI. Is there a way how to run specific columns through this filter before the export? I would expect it to work by default.

**Comment by billyaraujo** [comment]
Yes, I have also noticed the same. Any developments on this?

**Comment by Mastanka** [comment]
+1

**Comment by AgDude** [comment]
This is something I have talked about implementing, but haven't gotten to it (yet).  In the mean time, here is a workaround for customer filtering. It doesn't apply your cellFilter values, but it lets you override the default formatting.

```
  $provide.decorator('uiGridExporterService', function($delegate, $filter){
    $delegate.formatFieldAsCsv = function (field) {
      if (field.value == null) { // we want to catch anything null-ish, hence just == not ===
        return '';
      }
      if (typeof(field.value) === 'number') {
        return field.value;
      }
      if (typeof(field.value) === 'boolean') {
        return (field.value ? 'TRUE' : 'FALSE') ;
      }
      if (typeof(field.value) === 'string') {
        return field.value;
      }
      if ( field.value instanceof Date){
        return $filter('date')(field.value, 'yyyyMMdd');
      }

      return JSON.stringify(field.value);
    };

    return $delegate
  });
```

**Comment by JonathanPinney** [comment]
Anyone using this decorator with text values which may have commas will want to surround text fields with quotes.

```javascript
...
if (typeof (field.value) === 'string') { 
  return '"' + field.value + '"'; 
}
...

**Comment by sachetgupta** [comment]
If you have applied filter to columnDefs, then you just have to write a callback method which will execute that same filter while exporting the data.

Below is my proposed solution. This method would work for multiple filter with multiple parameters.

Column defination:
````
var rptColumnDefs = [ {
        field: 'field1',
        cellFilter: "filter1:'param1' | filter2 | filter3 : 4 | filter4 : 4 : 'param2'",
        displayName: "Field 1"  
    }, field: 'field2',
        cellFilter: "filter1:'param1' | filter2 | filter3 : 4 | filter4 : 4 : 'param2'",
        displayName: "Field 2"  
    }]
````

Exporter Callback function:

````
$scope.reportGridOptions = {
    enableFiltering: true,
    enableHorizontalScrollbar: uiGridConstants.scrollbars.NEVER,
    enableVerticalScrollbar: uiGridConstants.scrollbars.WHEN_NEEDED,
    exporterCsvFilename: '',
    columnDefs: rptColumnDefs,    
    exporterFieldCallback: function (grid, row, col, input) {
            if (col.cellFilter) { // check if any filter is applied on the column
                var filters = col.cellFilter.split('|'); // get all the filters applied
                angular.forEach(filters, function (filter) {
                    var filterName = filter.split(':')[0]; // fetch filter name
                    var filterParams = filter.split(':').splice(1); //fetch all the filter parameters
                    filterParams.unshift(input); // insert the input element to the filter parameters list
                    var filterFn = $filter(filterName); // filter function
                    // call the filter, with multiple filter parameters. 
                    //'Apply' will call the function and pass the array elements as individual parameters to that function.
                    input = filterFn.apply(this, filterParams); 
                })
                return input;
            }
            else
                return input;
        }
````

This generic function will help export the data as it is displayed in UI Grid (after applying the filters.)
Hope this helps someone.

**Comment by sachetgupta** [comment]
If you have applied filter to columnDefs, then you just have to write a callback method which will execute that same filter while exporting the data.

Below is my proposed solution. This method would work for multiple filter with multiple parameters.

Column defination:

```
var rptColumnDefs = [ {
        field: 'field1',
        cellFilter: "filter1:'param1' | filter2 | filter3 : 4 | filter4 : 4 : 'param2'",
        displayName: "Field 1"  
    }, field: 'field2',
        cellFilter: "filter1:'param1' | filter2 | filter3 : 4 | filter4 : 4 : 'param2'",
        displayName: "Field 2"  
    }]
```
Exporter Callback function:

```
$scope.reportGridOptions = {
    enableFiltering: true,
    enableHorizontalScrollbar: uiGridConstants.scrollbars.NEVER,
    enableVerticalScrollbar: uiGridConstants.scrollbars.WHEN_NEEDED,
    exporterCsvFilename: '',
    columnDefs: rptColumnDefs,    
    exporterFieldCallback: function (grid, row, col, input) {
            if (col.cellFilter) { // check if any filter is applied on the column
                var filters = col.cellFilter.split('|'); // get all the filters applied
                angular.forEach(filters, function (filter) {
                    var filterName = filter.split(':')[0]; // fetch filter name
                    var filterParams = filter.split(':').splice(1); //fetch all the filter parameters
                    filterParams.unshift(input); // insert the input element to the filter parameters list
                    var filterFn = $filter(filterName); // filter function
                    // call the filter, with multiple filter parameters. 
                    //'Apply' will call the function and pass the array elements as individual parameters to that function.
                    input = filterFn.apply(this, filterParams); 
                })
                return input;
            }
            else
                return input;
        }
}
```
This generic function will help export the data as it is displayed in UI Grid (after applying the filters.)
Hope this helps someone

**Comment by ComeOutToPlay** [comment]
I use a method to configure exportation that includes the requested scenary.
Here we use Excel, CSV and PDF exportation. We have a custom CSV and PDF.

Hope it helps!

`      
      factory.configureExporter = function (gridApi)

     {
        // Configuration done with gridOptions when initView() is run
        if (gridApi.grid == null || gridApi.grid.options == null)
        {
            return;
        }
        var gridOptions = gridApi.grid.options;
        gridOptions.enableGridMenu = true,
        gridOptions.exporterMenuCsv = false,
        gridOptions.exporterMenuPdf = false,
        
        // Format the DATE in case you use a cellTemplate
        // http://ui-grid.info/docs/#/api/ui.grid.exporter.api:GridOptions 
        gridOptions.exporterFieldCallback = function ( grid, row, col, value ){
          if ( col.colDef.cellTemplate == 'ui-grid/date-cell' || col.colDef.type == 'date' ){
            value = $filter('date')(value, 'yyyy-MM-dd');
          }
          return value;
        }
        gridOptions.gridMenuCustomItems = [
            {
                title: $filter('translate')('EXPORT_ALL_DATA_AS_EXCEL'),
                action: function ($event) {
                    
                    var sqlQuery = "SELECT ";
                    var list = [];

                    // Build columns to get from the query
                    $.each(gridOptions.columnDefs, function(index, column) {

                        if (column.field != null && column.name != null && column.name.trim() != "")
                        {
                            if ($.inArray(column.name, gridOptions.exporterSuppressColumns) == -1)
                            {
                                if (column.cellFilter != null && column.cellFilter != "")
                                {
                                    list.push('[' + column.field + '_fromCellFilter] AS ' + '[' + column.name + ']');   
                                }
                                else if (column.cellTemplate == 'ui-grid/date-cell' || column.type == 'date' )
                                {
                                    list.push('[' + column.field + '_formattedDate] AS ' + '[' + column.name + ']');   
                                }
                                else
                                {
                                    list.push('[' + column.field + '] AS ' + '[' + column.name + ']');
                                }
                            }
                        }
                    });

                    var rows = this.grid.rows;
                    var columns = this.grid.columns;

                    // Update cells that use cellFilter or cellTemplate
                    $.each(rows, function(index, row) {

                         $.each(columns, function(index, gridCol) {

                            if (gridCol.cellFilter != null && gridCol.cellFilter != "")
                            {
                                row.entity[gridCol.field + '_fromCellFilter' ] = gridApi.grid.getCellDisplayValue( row, gridCol );
                            }
                            else
                            {
                              if (gridCol.colDef.cellTemplate == 'ui-grid/date-cell' || gridCol.colDef.type == 'date' ){
                                row.entity[gridCol.field + '_formattedDate' ] = $filter('date')(row.entity[gridCol.field], 'yyyy-MM-dd');
                              }
                            }                            
                         });
                    })

                    sqlQuery += list.join(', ');

                    sqlQuery +=' INTO XLSX("download.xlsx",{headers:true}) FROM ?';
        
                    // At runtime we use grid.options.data
                    alasql(sqlQuery, [this.grid.options.data]);
                },
                order: 0
            }, 
            {
                title: $filter('translate')('EXPORT_ALL_DATA_AS_CSV'),
                action: function ($event) {

                    var exportColumnHeaders = uiGridExporterService.getColumnHeaders(this.grid, uiGridExporterConstants.ALL);                    
                    var exportData = uiGridExporterService.getData(this.grid, uiGridExporterConstants.ALL, uiGridExporterConstants.ALL, true);
                    var csvContent = uiGridExporterService.formatAsCsv(exportColumnHeaders, exportData, this.grid.options.exporterCsvColumnSeparator);
                    uiGridExporterService.downloadFile (this.grid.options.exporterCsvFilename, csvContent, this.grid.options.exporterOlderExcelCompatibility);

               },
                order:1
            },
            {
                title: $filter('translate')('EXPORT_ALL_DATA_AS_PDF'),
                action: function ($event) {

                    var exportColumnHeaders = uiGridExporterService.getColumnHeaders(this.grid, uiGridExporterConstants.ALL);
                    var exportData = uiGridExporterService.getData(this.grid, uiGridExporterConstants.ALL, uiGridExporterConstants.ALL, true);
                    var docDefinition = uiGridExporterService.prepareAsPdf(this.grid, exportColumnHeaders, exportData);

                    if (uiGridExporterService.isIE() || navigator.appVersion.indexOf("Edge") !== -1) {
                      uiGridExporterService.downloadPDF(this.grid.options.exporterPdfFilename, docDefinition);
                    } else {
                      pdfMake.createPdf(docDefinition).download();
                    }
                },
                order: 2
            }
        ];
    }
`

**Comment by mportuga** [comment]
This issue was addressed on version 4.4.6:
https://github.com/angular-ui/ui-grid/blob/master/CHANGELOG.md
